### PR TITLE
AUTO-646 - Cloudwatch alerting on NAT instances

### PR DESCRIPTION
**Pull request by snayler0**
Build Passing: https://gaautobots.atlassian.net/builds/browse/AM-AUT43/latest

### CODE REVIEW

File: amazonia/classes/single_instance.py [hunk 1]
```diff
@@ -85,6 +88,23 @@ def __init__(self, title, vpc, template, keypair, si_image_id, si_instance_type,
                                UserData=Base64(userdata)
                                ))
 
+        if is_nat and alert and alert_emails is not None:
+
```
Comment (ga-autobots) [inline amazonia/classes/single_instance.py]: blank line

File: amazonia/classes/single_instance.py [hunk 2]
```diff
@@ -85,6 +88,23 @@ def __init__(self, title, vpc, template, keypair, si_image_id, si_instance_type,
                                UserData=Base64(userdata)
                                ))
 
+        if is_nat and alert and alert_emails is not None:
```
Comment (ga-autobots) [inline amazonia/classes/single_instance.py]: 'is not none' not needed?

File: amazonia/classes/single_instance.py [hunk 3]
```diff
@@ -85,6 +88,23 @@ def __init__(self, title, vpc, template, keypair, si_image_id, si_instance_type,
...
+
+            snsName = title + 'topic'
+            self.topic = SNS(title, self.template, snsName, snsName)
+
+            for email in alert_emails:
+                self.topic.add_subscription(email, 'email')
+
+            metric = 'CPUUtilization'
+            self.topic.add_alarm(
+                'Alarms when {0} metric {1} reaches {2}'.format(self.single.title, metric, '60'),
```
Comment (ga-autobots) [inline amazonia/classes/single_instance.py]: use keyword arguments

File: amazonia/classes/sns.py [hunk 4]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
         :param display_name: The SNS display name
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: add AWS/Troposphere headers to docstrings

File: amazonia/classes/sns.py [hunk 5]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
...
         """
         title = unit_title + 'sns'
-        self.sns_topic = template.add_resource(Topic(title, TopicName=topic_name, DisplayName=display_name))
 
-        template.add_output(Output(
+        self.template = template
+        self.sns_topic = self.template.add_resource(Topic(title, TopicName=topic_name, DisplayName=display_name))
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: is TopicName needed here?

File: amazonia/classes/sns.py [hunk 6]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
...
+
+        self.template.add_output(Output(
             title,
             Value=Ref(self.sns_topic),
-            Description=Join('', ['SNS topic, created with Amazonia as part of ',
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: Look into reinstating Join here?

File: amazonia/classes/sns.py [hunk 7]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
...
+        using the add_subscription function above.
+        :param description: A description for the Alarm being created
+        :param metric: The metric to track and alarm on.
+        :param namespace: the namespace that the provided metric belongs to.
+        :param threshold: The threshold to alarm on
+        :param instance: an instance to refer to in particular
+        """
+
+        self.alarms.append(
+            self.template.add_resource(
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: do Subscriptions the same as here?

File: amazonia/classes/sns.py [hunk 8]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
...
+                cloudwatch.Alarm(
+                    '{0}Alarm{1}'.format(self.sns_topic.title, str(len(self.alarms))),
+                    AlarmDescription=str(description),
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: no need to str this

File: amazonia/classes/sns.py [hunk 9]
```diff
@@ -15,12 +15,60 @@ def __init__(self, unit_title, template, topic_name, display_name):
...
+                    AlarmActions=[Ref(self.sns_topic.title)],
+                    OKActions=[Ref(self.sns_topic.title)],
```
Comment (ga-autobots) [inline amazonia/classes/sns.py]: does this need .title

File: amazonia/schema.yaml [hunk 10]
```diff
@@ -370,4 +372,14 @@ db_hdd_size: # the size for the rds hard drive/allocated storage
   type:
     - 'number'
     - 'string'
-  nullable: False
\ No newline at end of file
+  nullable: False
+owner_emails:  # A list of email addresses to specify ownership of this stack and also send alerts to if
+  type: 'list' # alerts have been set up.
+  nullable: True
```
Comment (ga-autobots) [inline amazonia/schema.yaml]: Does this work with no emails?

File: test/sys_tests/test_sys_sns.py [hunk 11]
```diff
@@ -8,10 +8,45 @@ def main():
...
+        SourceDestCheck=True,
+        DependsOn=internet_gateway.title
+        ))
+
+    mysns = SNS(unit_title='test',
+                template=template,
+                topic_name='topic_name',
+                display_name='display_name')
+
+    mysns.add_subscription('some.email@some.domain.com', 'email')
```
Comment (ga-autobots) [inline test/sys_tests/test_sys_sns.py]: remove function unit tests here

File: test/unit_tests/test_single_instance.py [hunk 12]
```diff
@@ -55,7 +55,42 @@ def test_jump_with_hostedzone_creates_r53_record():
...
+
+    title = 'natreport'
+
+    si = create_si(title, is_nat=True, alert=True, alert_emails=['some@email.com'])
+
+    # Topic Created?
+    assert_equals(si.topic.sns_topic.title, title + 'sns')
+    assert_equals(si.topic.sns_topic.TopicName, title + 'topic')
+    assert_equals(si.topic.sns_topic.DisplayName, title + 'topic')
+    assert_in(si.template.outputs[title + 'sns'].Description, 'SNS topic created with Amazonia')
```
Comment (ga-autobots) [inline test/unit_tests/test_single_instance.py]: Update this after 'Join' is fixed

File: test/unit_tests/test_single_instance.py [hunk 13]
```diff
@@ -55,7 +55,42 @@ def test_jump_with_hostedzone_creates_r53_record():
...
+
+    # Subscription Created?
+    assert_equals(si.topic.sns_topic.Subscription[0].title, title + 'snsSubscription0')
+    assert_equals(si.topic.sns_topic.Subscription[0].Endpoint, 'some@email.com')
+    assert_equals(si.topic.sns_topic.Subscription[0].Protocol, 'email')
+
+    # Alarm Created?
+    assert_equals(si.topic.alarms[0].title, title + 'snsAlarm0')
+    assert_equals(si.topic.alarms[0].AlarmDescription, 'Alarms when ' + title + ' metric CPUUtilization reaches 60')
+    assert_equals(type(si.topic.alarms[0].AlarmActions[0]), type(Ref('abc')))
```
Comment (ga-autobots) [inline test/unit_tests/test_single_instance.py]: remove 'abc?'

File: test/unit_tests/test_single_instance.py [hunk 14]
```diff
@@ -55,7 +55,42 @@ def test_jump_with_hostedzone_creates_r53_record():
...
+
+    # Topic Created?
+    assert_equals(si.topic.sns_topic.title, title + 'sns')
+    assert_equals(si.topic.sns_topic.TopicName, title + 'topic')
+    assert_equals(si.topic.sns_topic.DisplayName, title + 'topic')
+    assert_in(si.template.outputs[title + 'sns'].Description, 'SNS topic created with Amazonia')
+
+    # Subscription Created?
+    assert_equals(si.topic.sns_topic.Subscription[0].title, title + 'snsSubscription0')
+    assert_equals(si.topic.sns_topic.Subscription[0].Endpoint, 'some@email.com')
```
Comment (ga-autobots) [inline test/unit_tests/test_single_instance.py]: Break these into 3 tests?

File: amazonia/schema.yaml (hunk 10 above)
Comment (snayler0) [inline amazonia/schema.yaml]: Yes, confirmed :)

File: amazonia/classes/sns.py (hunk 4 above)
Comment (snayler0) [inline amazonia/classes/sns.py]: done :)

**Comment by fuddstar** [comment]
:shipit:
Approved
:+1:

**Pull request merged**